File: 0001-0500/0202_isHappy.py
class Solution:
    def isHappy(self, n: int) -> bool:
        Set = set()
        while n not in Set:
            Set.add(n)
            num = str(n)
            x = 0
            for i in num:
                x += int(i) ** 2
            n = x
        return n == 1

# Alternative solution

class Solution:
    def isHappy(self, n: int) -> bool:
        fast = slow = n
        while fast != 1:
            fast = self.digitSquareSum(self.digitSquareSum(fast))
            if fast == 1:
                break
            slow = self.digitSquareSum(slow)
            if fast == slow:
                return False
        return True
            
    def digitSquareSum(self, n):
        res = 0
        while n:
            x = n % 10
            res += x ** 2
            n //= 10
        return res

File: 0001-0500/0202_isHappy.java
class Solution {
    public boolean isHappy(int n) {
        int fast = n;
        int slow = n;

        while (fast != 1) {
            slow = convertNumber(slow);
            fast = convertNumber(convertNumber(fast));
            if (fast != 1 && slow == fast) {
                return false;
            }
        }

        return true;
    }

    public int convertNumber(int n) {
        int result = 0;

        while (n > 0) {
            int mod = n % 10;
            result += Math.pow(mod, 2);
            n /= 10;
        }

        return result;
    }
}
